File: server/src/com/litbikes/model/Player.java
package com.litbikes.model;

import com.litbikes.dto.PlayerDto;
import com.litbikes.model.PowerUp.PowerUpType;

public class Player implements ICollidable {
	public enum PlayerEffect {
		NONE,
		SLOWED
	}

	protected int pid;
	protected Bike bike;
	private String name;
	private boolean crashed = false;
	private boolean spectating = false;
	private ICollidable crashedInto = null;
	private boolean isHuman;
	private PowerUpType currentPowerUpType = null;
	private PlayerEffect effect = PlayerEffect.NONE;
	
	public Player(int _pid, boolean _isHuman) {
		pid = _pid;
		isHuman = _isHuman;
	}
	
	public int getId() {
		return pid;
	}
	
	public String getName() {
		return name;
	}

	public Bike getBike() {
		return bike;
	}	
	
	public void setBike(Bike bike) {
		this.bike = bike;
	}

	public void setName(String name) {
		this.name = name;
	}

	public boolean isAlive() {
		return !crashed && !spectating;
	}
	
	public boolean isHuman() {
		return isHuman;
	}
	
	public PowerUpType getCurrentPowerUpType() {
		return currentPowerUpType;
	}
	
	public void setCurrentPowerUpType(PowerUpType type) {
		currentPowerUpType = type;
	}
	
	public PlayerEffect getEffect() {
		return effect;
	}
	
	public void setEffect(PlayerEffect effect) {
		this.effect = effect;
	}
	
	public PlayerDto getDto() {
		PlayerDto dto = new PlayerDto();
		dto.pid = pid;
		dto.name = name;
		dto.bike = bike.getDto();
		dto.crashed = crashed;
		dto.crashedInto = crashedInto != null ? crashedInto.getId() : null;
		dto.crashedIntoName = crashedInto != null ? crashedInto.getName() : null;
		dto.spectating = spectating;
		dto.currentPowerUp = currentPowerUpType;
		dto.effect = effect;
		return dto;
	}
	
	public void update() {
		if (isAlive()) {
			bike.updatePosition();	
		}
	}

	public void updateBike(Bike _bike) {
		// this is terrible
		bike.setPos(_bike.getPos());
		bike.setDir(_bike.getDir());
	}
	
	public boolean crashedIntoSelf() {
		if (crashedInto == null)
			return false;
		return crashedInto.getId() == pid;
	}	
	
	public boolean isSpectating() {
		return spectating;
	}

	public void setSpectating(boolean spectating) {
		this.spectating = spectating;
	}

	public ICollidable getCrashedInto() {
		return crashedInto;
	}

	public void setCrashedInto(ICollidable crashedInto) {
		this.crashedInto = crashedInto;
	}

	public boolean isCrashed() {
		return crashed;
	}

	public void setCrashed(boolean crashed) {
		this.crashed = crashed;
	}

	public void crashed(ICollidable collidedWith) {
		bike.crash();
		setCrashed(true);
		setCrashedInto(collidedWith);
		setSpectating(true);
	}
	
	public void respawn(Spawn spawn) {
		bike.init(spawn, false);
		setCrashed(false);
		setCrashedInto(null);
		setSpectating(false);
		effect = PlayerEffect.NONE;
	}
}

File: client/src/model/player.ts
module Model {
    import PlayerDto = Dto.PlayerDto;
    export class Player {        
        private pid: number;
        private name: string;
        private bike: Bike;
        private crashed: boolean;
        private spectating: boolean;
        private deathTimestamp: number;
        private crashedInto: number; // pid last crashed into
        private crashedIntoName: string;
        private score: number;
        private isControlledPlayer: boolean;
        private currentPowerUp: string;
        private effect: string;

        constructor(pid: number, name: string, bike: Bike, crashed: boolean, 
            spectating: boolean, deathTimestamp: number, crashedInto: number, 
            crashedIntoName: string, score: number, isControlledPlayer: boolean) {
            this.pid = pid;
            this.name = name;
            this.bike = bike;
            this.crashed = crashed;
            this.spectating = spectating;
            this.deathTimestamp = deathTimestamp;
            this.crashedInto = crashedInto;
            this.crashedIntoName  = crashedIntoName;
            this.score = score;
            this.isControlledPlayer = isControlledPlayer;
            
            if (this.isAlive) {
                this.bike.respawned();
            }
        }

        public getPid(): number {
            return this.pid;
        }
        public getName(): string {
            return this.name;
        }
        public setName(name: string) {
            this.name = name;
        }
        public getBike(): Bike {
            return this.bike;
        }
        public setBike(bike: Bike) {
            this.bike = bike;
        }
        public isCrashed() : boolean {
            return this.crashed;
        }
        public getCrashedInto() : number {
            return this.crashedInto;
        }
        public getCrashedIntoName() : string {
            return this.crashedIntoName;
        }
        public isAlive(): boolean {
            return !this.spectating && !this.crashed;
        }
        private isVisible() : boolean {
            return !this.spectating || this.bike.isCrashing();
        }
        public isSpectating() : boolean {
            return this.spectating;
        }
        public getCurrentPowerUp(): string {
            return this.currentPowerUp;
        }
        public getEffect(): string {
            return this.effect || "none";
        }

        public update() {
            this.bike.update(this.isAlive());
        }

        public draw(p: p5, showName: boolean) {
            if (this.isVisible()) {
                let showRespawnRing = this.isAlive() && this.isControlledPlayer;
                this.bike.draw(p, showRespawnRing, this.isControlledPlayer);
                
                if (showName) {
                    p.textSize(15);
                    p.textAlign('center', 'middle');
                    p.text(this.name, this.bike.getPos().x, Math.max(0, this.bike.getPos().y - 15));
                }
                
                //p.text(this.effect ? this.effect : "none", this.bike.getPos().x, Math.max(0, this.bike.getPos().y - 15));

            }
        }

        public updateFromDto( dto : PlayerDto ) {
            let wasAlive = this.isAlive();

            if ( !this.crashed && dto.crashed ) {
                this.bike.crash();
                this.deathTimestamp = dto.deathTimestamp || Math.floor(Date.now());
            }

            if (!dto.crashed && (this.crashed || (this.spectating && !dto.spectating))) {
                this.bike.respawned();
            }

            let oldPowerUp = this.currentPowerUp;            
            this.crashed = dto.crashed;
            this.crashedInto = dto.crashedInto;
            this.crashedIntoName = dto.crashedIntoName;
            this.spectating = dto.spectating;
            this.score = dto.score;
            this.currentPowerUp = dto.currentPowerUp ? dto.currentPowerUp.toLowerCase() : null;
            this.effect = dto.effect;
            this.bike.updateFromDto(dto.bike);
          
        }
    }
}